pageserver/src/page_service.rs

```diff
@@ -899,165 +995,383 @@ impl PageServerHandler {
+
+        //
+        // Create Batching future.
+        //
 
-                // Map handler result to protocol behavior.
-                // Some handler errors cause exit from pagestream protocol.
-                // Other handler errors are sent back as an error message and we stay in pagestream protocol.
-                for handler_result in handler_results {
-                    let response_msg = match handler_result {
-                        Err(e) => match &e {
-                            PageStreamError::Shutdown => {
-                                // If we fail to fulfil a request during shutdown, which may be _because_ of
-                                // shutdown, then do not send the error to the client.  Instead just drop the
-                                // connection.
-                                span.in_scope(|| info!("dropping connection due to shutdown"));
-                                return Err(QueryError::Shutdown);
+        enum BatchState {
+            Building(Option<Box<BatchedFeMessage>>),
+            ReadMessagesEnded(Option<Box<BatchedFeMessage>>),
+            ExecutorEnded,
+        }
+        let (batch_tx, mut batch_rx) = tokio::sync::watch::channel(Arc::new(
+            std::sync::Mutex::new(BatchState::Building(None)),
+        ));
+        let notify_batcher = Arc::new(tokio::sync::Notify::new());
+        let batcher = {
+            let notify_batcher = notify_batcher.clone();
+            async move {
+                scopeguard::defer! {
+                    debug!("exiting");
+                }
+                'outer: loop {
+                    let maybe_req = requests_rx.recv().await;
+                    let Some(req) = maybe_req else {
+                        batch_tx.send_modify(|pending_batch| {
+                            let mut guard = pending_batch.lock().unwrap();
+                            match &mut *guard {
+                                BatchState::Building(batch) => {
+                                    *guard = BatchState::ReadMessagesEnded(batch.take());
+                                }
+                                BatchState::ReadMessagesEnded(_) => {
+                                    unreachable!("we exit the first time")
+                                }
+                                BatchState::ExecutorEnded => {
+                                    debug!("observing executor ended when reading upstream");
+                                }
                             }
-                            PageStreamError::Reconnect(reason) => {
-                                span.in_scope(|| info!("handler requested reconnect: {reason}"));
-                                return Err(QueryError::Reconnect);
+                        });
+                        break;
+                    };
+                    // don't read new requests before this one has been processed
+                    let mut req = Some(req);
```

Comment:
nit: this is never `None` outside the loop below, right? Seems that we use `Option` here just for the `take` and `replace` interface.